### PR TITLE
Add `ADLSCreateObjectOperator`

### PR DESCRIPTION
**Pull request by dabla**
Hello, as discussed in the previous [PR](https://github.com/apache/airflow/pull/37103) I created this new pull request for introducing the DataToADLSOperator.  This operator allows you to use the AzureDataLakeStorageV2Hook without the need to write custom python code to interact with the Hook. 

There is already a LocalFilesystemToADLSOperator which uses the older AzureDataLakeHook, but that operator doesn't allow you to directly store an XCom to Fabric, you need to create a local file first before uploading it, which in this case would mean creating an additional task to do it while in fact it's not necessary as the AzureDataLakeStorageV2Hook allows you to achieve the same without that intermediate step.

That's why I introduced this DataToADLSOperator, we already use it at our company as a custom operator and it works great as this easily allows you to upload an XCom to Fabric, without the need to write python code interacting with the AzureDataLakeStorageV2Hook.  This means a more consice DAG, less code which improves readability and reduces complexity and of course makes it easy to use.

Here is a simplified example on how to use the operator:

```python
from airflow import DAG
from airflow.operators.jira_plugin import JiraOperator
from airflow.providers.microsoft.azure.transfers.data_to_local import DataToADLSOperator
from datetime import datetime

default_args = {
    'owner': 'airflow',
    'start_date': datetime(2024, 1, 1),
}

with DAG('jira_example_dag', default_args=default_args, schedule_interval=None) as dag:
    create_issue_task = JiraOperator(
        task_id='create_issue',
        jira_conn_id='jira_connection',
        method='POST',
        endpoint='issue',
        json={
            'fields': {
                'project': {'key': 'PROJ'},
                'summary': 'This is a test issue created from Airflow',
                'description': 'This is the description of the test issue.',
                'issuetype': {'name': 'Task'},
            }
        },
	dag=dag,
    )

    get_issue_task = JiraOperator(
        task_id='get_issue',
        jira_conn_id='jira_connection',
        method='GET',
        endpoint='issue/AIRFLOW-123',
	dag=dag,
    )

    write_to_azure_task = DataToADLSOperator(
        task_id="write_to_azure",
	azure_data_lake_conn_id="ms_azure_data_lake_connection",
	file_system_name="JIRA Monitoring",
	file_name="Files/JIRA/issues/Airflow/AIRFLOW-123.json",
	data=get_issue_task.output,
	overwrite=True,
	dag=dag,
    )

    create_issue_task >> get_issue_task >> write_to_azure_task
```

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by Joffreybvn — APPROVED** [review]

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py [hunk 1]
```diff
@@ -96,3 +96,53 @@ def execute(self, context: Context) -> None:
             blocksize=self.blocksize,
             **self.extra_upload_options,
         )
+
+
+class DataToADLSOperator(BaseOperator):
```
Comment (eladkal) [inline airflow/providers/microsoft/azure/transfers/local_to_adls.py]: I checked and we have the equivalent operator in S3

https://github.com/apache/airflow/blob/2bc103698fad1b7c0bebed0d5cbdda0ec7ea83c3/airflow/providers/amazon/aws/operators/s3.py#L352


So please lets rename the operator to be `ADLSCreateObjectOperator` and place it in `operators/adsl.py`
Lets also try to keep parameters name according to the conventions we already have in the S3 (where it make sense) for example `replace` rather than `overwrite` this is very helpful for users who work in multi cloud environment. Note also about the extra parameters there are in the S3 example maybe would be good to add them like `encryption`, `encoding`

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py (hunk 1 above)
Comment (dabla) [reply airflow/providers/microsoft/azure/transfers/local_to_adls.py]: Hello Elad thank you for the review good to know, I'll fix that.

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py [hunk 2]
```diff
@@ -59,22 +56,3 @@ def test_execute_raises_for_bad_glob_val(self, mock_hook):
...
-        operator.execute(None)
-        mock_hook.return_value.upload_file.assert_called_once_with(
-            local_path=LOCAL_PATH,
-            remote_path=REMOTE_PATH,
-            nthreads=64,
-            overwrite=True,
-            buffersize=4194304,
-            blocksize=4194304,
-            run=False,  # extra upload options
-        )
```
Comment (eladkal) [inline tests/providers/microsoft/azure/transfers/test_local_to_adls.py]: why do we remove this test if no changes were made in the `LocalFilesystemToADLSOperator` itself?

Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py]: Oeps that will be a wrong refactoring of my part :)

Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py]: I still have following error related to docs but dunno what I did wrong, as I only moved docs from local_to_adls transfers to adls in operators.  I always receive following error in static checks and when I compare the files I edited I don't see any issues:

```
apache-airflow-providers-microsoft-azure                     Traceback (most recent call last):
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/site-packages/sphinx/cmd/build.py", line 281, in build_main
  apache-airflow-providers-microsoft-azure                         app.build(args.force_all, args.filenames)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/site-packages/sphinx/application.py", line 347, in build
  apache-airflow-providers-microsoft-azure                         self.builder.build_update()
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 307, in build_update
  apache-airflow-providers-microsoft-azure                         self.build(['__all__'], to_build)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 326, in build
  apache-airflow-providers-microsoft-azure                         updated_docnames = set(self.read())
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/contextlib.py", line 120, in __exit__
  apache-airflow-providers-microsoft-azure                         next(self.gen)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/site-packages/sphinx/util/logging.py", line 218, in pending_warnings
  apache-airflow-providers-microsoft-azure                         memhandler.flushTo(logger)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/site-packages/sphinx/util/logging.py", line 183, in flushTo
  apache-airflow-providers-microsoft-azure                         logger.handle(record)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/logging/__init__.py", line 1599, in handle
  apache-airflow-providers-microsoft-azure                         self.callHandlers(record)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/logging/__init__.py", line 1661, in callHandlers
  apache-airflow-providers-microsoft-azure                         hdlr.handle(record)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/logging/__init__.py", line 950, in handle
  apache-airflow-providers-microsoft-azure                         rv = self.filter(record)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/logging/__init__.py", line 811, in filter
  apache-airflow-providers-microsoft-azure                         result = f.filter(record)
  apache-airflow-providers-microsoft-azure                       File "/usr/local/lib/python3.8/site-packages/sphinx/util/logging.py", line 426, in filter
  apache-airflow-providers-microsoft-azure                         raise exc
  apache-airflow-providers-microsoft-azure                     sphinx.errors.SphinxWarning: 
  /opt/airflow/docs/apache-airflow-providers-microsoft-azure/_api/airflow/providers/microsoft/azure/operators/adls/index.rst:17:Definition list ends without a blank line; unexpected 
  unindent.
  apache-airflow-providers-microsoft-azure                     
  apache-airflow-providers-microsoft-azure                     [91mWarning, treated as error:[39;49;00m
  apache-airflow-providers-microsoft-azure                     
  /opt/airflow/docs/apache-airflow-providers-microsoft-azure/_api/airflow/providers/microsoft/azure/operators/adls/index.rst:17:Definition list ends without a blank line; unexpected 
  unindent.
  apache-airflow-providers-microsoft-azure                    : ################################################################################
  
  
No packages to retry. No more passes are needed.
No packages to retry. No more passes are needed.
No packages to retry. No more passes are needed.
fThere were some build errors remaining.


############################## Start spelling errors summary ##############################

```

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py
Comment (eladkal) [inline tests/providers/microsoft/azure/transfers/test_local_to_adls.py:1]: There are still some unnecessary changes to this file.
Can you please revert them?

File: airflow/cli/commands/connection_command.py [hunk 3]
```diff
@@ -314,7 +314,7 @@ def connections_delete(args):
 def connections_import(args):
     """Import connections from a file."""
     if os.path.exists(args.file):
-        _import_helper(args.file, args.overwrite)
+        _import_helper(args.file, args.replace)
```
Comment (eladkal) [inline airflow/cli/commands/connection_command.py]: how is this change related to the PR?

Comment (dabla) [reply airflow/cli/commands/connection_command.py]: That's due to the refactoring I did, seems like Pycharm went a bit to far, didn't even catch that one, thanks for pointing it out

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py
Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:1]: Don't understand what's wrong here...

File: airflow/cli/commands/connection_command.py (hunk 3 above)
Comment (dabla) [reply airflow/cli/commands/connection_command.py]: I've reverted back that change

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py
Comment (eladkal) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:1]: There shouldn't be changes on this file as part of your PR.
You are removing lines from the code on this file

Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:1]: Ok now I understand, but that file was changed because the operator was tested there initially due to the fact that it was located in the transfers module.  As we moved the operator from transfers to operator module, I had to move the test also hence why it has changed.  Still I don't understand why the docs build section is failing as I just also moved the docs (and even improved them, found even an error in existing one for adls operator) from transfers to operator also.

Comment (eladkal) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:1]: Sorry I don't follow let me comment specifically on the lines

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py [hunk 4]
```diff
@@ -17,7 +17,6 @@
 # under the License.
 from __future__ import annotations
 
-import json
```
Comment (eladkal) [inline tests/providers/microsoft/azure/transfers/test_local_to_adls.py:20]: why? https://github.com/apache/airflow/pull/37821/files#r1514286502

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py [hunk 5]
```diff
@@ -28,11 +27,9 @@
 )
 
 TASK_ID = "test-adls-upload-operator"
-FILE_SYSTEM_NAME = "Fabric"
```
Comment (eladkal) [inline tests/providers/microsoft/azure/transfers/test_local_to_adls.py:31]: why? https://github.com/apache/airflow/pull/37821/files#r1514286502

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py [hunk 6]
```diff
@@ -28,11 +27,9 @@
...
 LOCAL_PATH = "test/*"
 BAD_LOCAL_PATH = "test/**"
 REMOTE_PATH = "TEST-DIR"
-DATA = json.dumps({"name": "David", "surname": "Blain", "gender": "M"}).encode("utf-8")
```
Comment (eladkal) [inline tests/providers/microsoft/azure/transfers/test_local_to_adls.py:35]: why? https://github.com/apache/airflow/pull/37821/files#r1514286502

Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:35]: Same reason, test moved from location, this line was added to test the new operator and is not used by the operator being tested here, it is now in the test file (tests/providers/microsoft/azure/operators/test_adls_create.py) of the new operator:

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py (hunk 5 above)
Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:31]: Same reason, test moved from location, this line was added to test the new operator and is not used by the operator being tested here, it is now in the test file (tests/providers/microsoft/azure/operators/test_adls_create.py) of the new operator:

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py (hunk 4 above)
Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:20]: Same reason, test moved from location, this line was added to test the new operator and is not used by the operator being tested here, it is now in the test file (tests/providers/microsoft/azure/operators/test_adls_create.py) of the new operator:

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py (hunk 6 above)
Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:35]: If you compare this file with main you will see it is identical, those modifications are due to the fact that we moved the operator and thus also the tests associated to it, now the new operator has a dedicated test file

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py (hunk 5 above)
Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:31]: The test_local_to_adls.py is now exactly the same as in main

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py (hunk 6 above)
Comment (eladkal) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:35]: It is not.

https://github.com/apache/airflow/blob/354e633ddd7da5fb8310ef6bd1c5b9e5e8763848/tests/providers/microsoft/azure/transfers/test_local_to_adls.py#L35

this line exist in main. In this PR you are removing it.

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py (hunk 5 above)
Comment (eladkal) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:31]: no...
You are requesting to remove this line.
https://github.com/apache/airflow/blob/354e633ddd7da5fb8310ef6bd1c5b9e5e8763848/tests/providers/microsoft/azure/transfers/test_local_to_adls.py#L31

Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:31]: Indeed this line was also only used for testing new Operator, same logic as above, this is all due to moving the operator to another module (from transfers to operator), which meant moving tests also to corresponding operator module, so that constant ins't needed anymore here, its moved to test_adls_create.py now where the new operator is tested.

File: tests/providers/microsoft/azure/transfers/test_local_to_adls.py (hunk 6 above)
Comment (dabla) [reply tests/providers/microsoft/azure/transfers/test_local_to_adls.py:35]: Ok now I understand what you mean, sorry for that, will remove this file from the pull request indeed it doesn't have to be there as nothing has to be changed there, this is l due to the refactoring I did.

**Comment by dabla** [comment]
> @dabla Let me clarify...
> 
> Your PR is requesting to merge changes on 6 files. One of them is: `tests/providers/microsoft/azure/transfers/test_local_to_adls.py`
> 
> Screenshot 2024-03-07 at 10 01 16
> My point is that your PR should change only 5 files. There should be no changes on `tests/providers/microsoft/azure/transfers/test_local_to_adls.

Ok now I understand what you mean, sorry for that, will remove this file from the pull request indeed it doesn't have to be there as nothing has to be changed there, this is l due to the refactoring I did.

File: tests/system/providers/microsoft/azure/example_adls_list.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+from datetime import datetime
+
+from airflow import models
+from airflow.providers.microsoft.azure.operators.adls import ADLSListOperator
+
+LOCAL_FILE_PATH = os.environ.get("LOCAL_FILE_PATH", "localfile.txt")
+REMOTE_FILE_PATH = os.environ.get("REMOTE_LOCAL_PATH", "remote.txt")
+
+ENV_ID = os.environ.get("SYSTEM_TESTS_ENV_ID")
+DAG_ID = "example_adls_delete"
```
Comment (eladkal) [inline tests/system/providers/microsoft/azure/example_adls_list.py]:
```suggestion
DAG_ID = "example_adls_list"
```

**Comment by dabla** [comment]
> @dabla Let me clarify...
> 
> Your PR is requesting to merge changes on 6 files. One of them is: `tests/providers/microsoft/azure/transfers/test_local_to_adls.py`
> 
> Screenshot 2024-03-07 at 10 01 16
> My point is that your PR should change only 5 files. There should be no changes on `tests/providers/microsoft/azure/transfers/test_local_to_adls.py`

Ok it's because the branch was based on the other PR with proxy fix where I had to the new operator, so it will be hard to remove that file.  What do you want me to do, create a new PR for this one based on main?

**Comment by eladkal** [comment]
> Ok @eladkal it's because the branch was based on the other PR with proxy fix where I had to the new operator, so it will be hard to remove that file. What do you want me to do, create a new PR for this one based on main?

It's just 3 lines. It would be simpler to just re-add these lines to the file (they exist in main but not in your branch) then just commit it. Once you do that the diff will be no changes to the file.

Feel free to send me message on Slack if you need help

**Comment by dabla** [comment]
> > Ok @eladkal it's because the branch was based on the other PR with proxy fix where I had to the new operator, so it will be hard to remove that file. What do you want me to do, create a new PR for this one based on main?
> 
> It's just 3 lines. It would be simpler to just re-add these lines to the file (they exist in main but not in your branch) then just commit it. Once you do that the diff will be no changes to the file.
> 
> Feel free to send me message on Slack if you need help

Hello Edal, first of all thank you for your patience now I fully understand what happend.  This file is indeed not the same as the main branch (I forgot to sync main on my repo hence why I didn't see the latestes updates and diff).  But it should be changed, because it contains remains of the test of that new operator I had to remove back then if you remember, but on that PR I apparently forgot to remove those unused constants.  Those constants are not used in that test anymore, so I would keep the changes as they clean up the test, but if you prefer to keep it then I will re-add those 3 lines if you want.  Let me know what you I have to do.

**Comment by eladkal** [comment]
OK now i get it.
the claim is that the changes were added in https://github.com/apache/airflow/pull/37103 by mistake :) 
This is why separating unrelated changes in PRs is so important it. It would have been much easier to understand and review if a separated PR would raise just to revert the unneeded changes (no need to do that now)

So all left here is just to fix the doc build.

**Comment by dabla** [comment]
> OK now i get it. the claim is that the changes were added in #37103 by mistake :) This is why separating unrelated changes in PRs is so important it. It would have been much easier to understand and review if a separated PR would raise just to revert the unneeded changes (no need to do that now)
> 
> So all left here is just to fix the doc build.

Exactly, it was my mistake as I didn't clean up the test well in my previous pull reuest hence all the confussion, so apologies once again.  Yeah the doc thing is a mysterie to me imho, some help or some pointers in a direction could help here :)

**Comment by eladkal** [comment]
> Exactly, it was my mistake as I didn't clean up the test well in my previous pull request hence all the confussion, so apologies once again. Yeah the doc thing is a mysterie to me imho, some help or some pointers in a direction could help here :)

hard to tell.
Can you extract the adsl list docs to a separated PR? (again, it's not part of the `ADLSCreateObjectOperator`)
That way we can check it separately and see which one is causing the doc build to fail

**Comment by dabla** [comment]
> > Exactly, it was my mistake as I didn't clean up the test well in my previous pull request hence all the confussion, so apologies once again. Yeah the doc thing is a mysterie to me imho, some help or some pointers in a direction could help here :)
> 
> hard to tell. Can you extract the adsl list docs to a separated PR? (again, it's not part of the `ADLSCreateObjectOperator`) That way we can check it separately and see which one is causing the doc build to fail

I reverted that change but the error stays the same

**Comment by dabla** [comment]
@eladkal I finally managed to make the build pass, I kept the example for the list operator as I found what was causing the issue ( a missing single quote :))

**Review by eladkal — APPROVED** [review]

**Comment by dabla** [comment]
Thanks @eladkal :)
